### ISSUE TITLE
MODL Syntax Versioning

### ISSUE THREAD

**Issue by twalmsley** [MODLanguage/modl.uk#11]
It is implied, but not specified, that the syntax version is a positive integer greater than zero, is this correct and can it be specified? (I don't think there's a need to support major.minor.patch version numbers such as 1.0.2)

How should the syntax version be handled? E.g. interpreters written for version N should reject MODL files with a version greater than N.

**Comment by elliottinvent** [comment]
I've added some clarity to answer the first point:
https://modl.uk/specification#interpreter-rules (number 4)

On the second point, the java-interpreter returns an error if the version number is higher than it is prepared for, e.g:

```
*V=2
test=1
```
 -> `Can't handle MODL version 2 `

I agree that we should specify how interpreters should handle versions higher than they are designed for and interested in your thoughts on this. Ideally they'd attempt interpretation and if succesful return a warning. If unsuccesful prepend any error message with something like "Interpreter not designed for MODL version x."

**Comment by twalmsley** [comment]
Its possible that a version 2 MODL file could in fact just be made up of version 1 constructs, so the _parser_ could handle it. I would say we go ahead and try, but if we get any kind of error we report it as something like `MODL Version 1 interpreter cannot process this MODL Version 2 file.`

**Comment by twalmsley** [comment]
Added tests for higher version numbers.
Interpreters should try to process the file, but if there's
an interpreter error report it as something like:
`MODL Version 1 interpreter cannot process this MODL Version 100 file.`

Parser syntax errors will prevent this because the error is thrown before the interpreter has had a chance to interpret the `*VERSION` command. We _could_ pre-process the MODL file to check for a `*VERSION` command before parsing but I prefer to let the lexer and parser do their job first. 

Fixed in the Ruby interpreter - Java and JS versions to follow later.

**Comment by elliottinvent** [comment]
Still need to make this clear in spec

**Comment by elliottinvent** [comment]
@twalmsley what's the latest thinking on this, so I can spec it please?

**Comment by twalmsley** [comment]
Looking back through the Slack discussion, the conclusion was that we're already doing the right thing but @alexdalitz didn't like the wording of the error - probably because it was masking the underlying error. 
I suggest we throw the underlying error and append a warning that the version number is too high for the current interpreter version, something like:
```
Interpreter Error: Numeric Keys Not Allowed. (MODL Version 2 not supported by this version 1 interpreter)
```

**Comment by alexdalitz** [comment]
That works for me

**Comment by twalmsley** [comment]
Fixed Ruby version - will be in release v0.3.14.
Java version to follow.

**Comment by twalmsley** [comment]
Java version updated - fix will be in v0.0.9
